### ISSUE TITLE
Do we like the pattern of returning Optionals from factories?

### ISSUE THREAD

**Issue by qualidafial** [jdbi/jdbi#676]
I know it's the new shiny in Java 8, and it makes sense to return `Optional` in lots of places. `Query.findFirst()` is a good example.

However when implementing our factory interfaces (specifically `ArgumentFactory`, `ColumnMapperFactory`, `RowMapperFactory`, and `SqlArrayTypeFactory`) it can get rather tedious.

I've wondered if we weren't better off with separate `accepts(...)` and `build(...)` methods like we used in 2.x.

I think with the recent addition of methods like `ColumnMapperFactory.of(Type, ColumnMapper)`, the benefit of having the factory class be a `@FunctionalInterface` is rather diminished.

@jdbi/contributors What do you all think? Should I do a spike and see how it looks?

**Comment by arteam** [comment]
We have a lot of `Optional`s for sure. I don't have any compelling arguments for any side. I think we could try this change and see whether it reduces complexity or not. Generally, I like the way factories implemented right now.

**Comment by stevenschlansker** [comment]
I'm similarly indifferent for the most part.  I often prefer the single method approach because sometimes you have to do some work to determine if you support an operation, and then you repeat that work in the `accepts` and `build` method.  If you want to spike a change I'd be curious to see it, but I'm happy with the state as is.

**Comment by qualidafial** [comment]
I've given this some thought, and I think you guys are right. The basis of this ticket was that I thought writing _correct_ factory implementations was tedious. However we've already got code in place too make it easy for end users to do the right thing:

* ArgumentFactory -> AbstractArgumentFactory
* ColumnMapperFactory -> ColumnMapperFactory.of(Type, ColumnMapper), Configurable.registerColumnMapper(Type, ColumnMapper)
* RowMapperFactory -> RowMapperFactory.of(Type, RowMapper), Configurable.registerRowMapper(Type, RowMapper)
* SqlArrayTypeFactory -> Configurable.register(Class, String sqlTypeName)

So I think we've got 99% of cases covered.

Closing.
